tests/functional/test_s3.py

```diff
@@ -225,6 +244,227 @@ def progress_callback(amount):
         self.assertEqual(self.progress, 55)
 
 
+class TestUploadFileobj(BaseTransferTest):
+    def setUp(self):
+        super(TestUploadFileobj, self).setUp()
+        self.contents = six.BytesIO(b'fooo')
+
+    def stub_put_object(self):
+        put_object_response = {
+            "ETag": self.etag,
+            "ResponseMetadata": {
+                "HTTPStatusCode": 200
+            }
+        }
+        expected_params = {
+            "Bucket": self.bucket,
+            "Key": self.key,
+            "Body": botocore.stub.ANY
+        }
+        self.stubber.add_response(
+            method='put_object', service_response=put_object_response,
+            expected_params=expected_params)
+
+    def stub_upload_part(self, part_number):
+        upload_part_response = {
+            'ETag': self.etag,
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200
+            }
+        }
+        expected_params = {
+            "Bucket": self.bucket,
+            "Key": self.key,
+            "Body": botocore.stub.ANY,
+            "PartNumber": part_number,
+            "UploadId": self.upload_id
+        }
+        self.stubber.add_response(
+            method='upload_part', service_response=upload_part_response,
+            expected_params=expected_params)
+
+    def stub_multipart_upload(self, num_parts):
+        self.stub_create_multipart_upload()
+
+        # Add the responses for each UploadPartCopy
+        parts = []
+        for i in range(num_parts):
+            # Fill in the parts
+            part_number = i + 1
+            self.stub_upload_part(part_number=part_number)
+            parts.append({'ETag': self.etag, 'PartNumber': part_number})
+
+        self.stub_complete_multipart_download(parts)
+
+    def test_client_upload(self):
+        self.stub_put_object()
+        with self.stubber:
+            # The stubber will assert that all the right parameters are called.
+            self.s3.meta.client.upload_fileobj(
```

Comment:
It would probably be good if we had an assertion for all of these. I am thinking just making sure every stubbed response in the stubber got consumed should be fine.